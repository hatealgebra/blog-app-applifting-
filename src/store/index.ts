import { configureStore, combineReducers } from '@reduxjs/toolkit';

import { persistStore, persistReducer } from 'redux-persist';
import storage from 'redux-persist/lib/storage';
import authReducer from './slices/auth.slices';
import articleFeedReducer from './slices/articleFeed.slice';
import adminReducer from './slices/admin.slices';

// REDUX-PERSIST

export const persistConfig = {
  key: 'root',
  storage,
};

export const persistedReducer = persistReducer(persistConfig, authReducer);

export const reducer = combineReducers({
  // FIXME: Just store the data where it is needed, use multiple stores
  articleFeed: articleFeedReducer,
  admin: adminReducer,
});

const store = configureStore({
  reducer: { reducer, persistedReducer },
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware({
      serializableCheck: false,
    }),
  devTools: true,
});

export const persistor = persistStore(store);
export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;

export default store;
