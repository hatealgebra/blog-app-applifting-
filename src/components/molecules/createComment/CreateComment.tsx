import React from 'react';
import { Components } from '@customTypes/declarations';
import publishComment from '../../../helpers/commenting.helper';
import { useAppSelector } from '../../../store/hooks';
import { selectAuthName } from '../../../store/slices/auth.slices';

import Avatar from '../../atoms/avatar/Avatar';
import Button from '../../atoms/button/Button';
import { ErrorText } from '../../atoms/errorText/error.styled';
import { StyledTextArea } from '../../atoms/input/input.styled';
import { StyledCreateCommentForm } from './createComment.styled';

export enum FormValidation {
  EMPTY = '* Comment is empty or too short. Min. length of the text should be 25 characters.',
  TOO_LONG = '* Comment is too long. Maximum is 250 characters.',
  PASSED = '',
}

interface CreateCommentProps {
  articleId: string;
  setComments: React.Dispatch<
    React.SetStateAction<Components['schemas']['Comment'][]>
  >;
}

// TODO: UseEffect for form handling
const CreateComment = ({ articleId, setComments }: CreateCommentProps) => {
  const [isActive, setIsActive] = React.useState(false);
  const [content, setContent] = React.useState('');
  const [formHandling, setFormHandling] = React.useState<FormValidation>(
    FormValidation.PASSED
  );
  const loggedUser = useAppSelector(selectAuthName);

  const onSubmit = (
    e: React.FormEvent,
    setCurrComments: React.Dispatch<
      React.SetStateAction<Components['schemas']['Comment'][]>
    >
  ) => {
    e.preventDefault();
    if (content.length < 25) {
      setFormHandling(FormValidation.EMPTY);
    } else if (content.length > 250) {
      setFormHandling(FormValidation.TOO_LONG);
    } else if (loggedUser) {
      publishComment(articleId, loggedUser, content, setCurrComments);
      setIsActive(false);
      setFormHandling(FormValidation.PASSED);
      setContent('');
    }
  };

  return (
    <StyledCreateCommentForm
      onSubmit={(e) => onSubmit(e, setComments)}
      onClick={() => {
        if (!loggedUser && typeof window !== `undefined`) {
          window.alert('You need to be signed in to comment this article!');
        }
      }}
      onBlur={(e: React.FocusEvent<HTMLFormElement>) =>
        e.relatedTarget === null && setIsActive(false)
      }
    >
      <Avatar />
      <StyledTextArea
        disabled={!loggedUser}
        onFocus={() => setIsActive(true)}
        placeholder="Join the discussion"
        rows={isActive ? 8 : 1}
        name="comment"
        value={content}
        onChange={(e) => {
          return setContent(e.target.value);
        }}
      />
      {isActive && <Button type="submit">Send comment</Button>}
      <ErrorText>{formHandling}</ErrorText>
    </StyledCreateCommentForm>
  );
};

export default CreateComment;
