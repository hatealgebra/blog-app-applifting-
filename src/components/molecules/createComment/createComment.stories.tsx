import { expect, jest } from '@storybook/jest';
import React, { useEffect, useState } from 'react';
import { Provider } from 'react-redux';
import { userEvent, within } from '@storybook/testing-library';
import store from '../../../store';
import CreateComment from './CreateComment';

const CreateCommentTemplate = (args) => {
  const [, setCurrComments] = useState([]);

  useEffect(() => {
    args.setComments();
  }, [args]);

  return (
    <Provider store={store}>
      <CreateComment articleId={args.articleId} setComments={setCurrComments} />
    </Provider>
  );
};

export const CreateCommentExample = CreateCommentTemplate.bind({});
CreateCommentExample.args = {
  articleId: '1',
  setComments: jest.fn(() => []),
};

export const CannotCreateComment = CreateCommentTemplate.bind({});
CannotCreateComment.args = CreateCommentExample.args;
CannotCreateComment.play = async ({ canvasElement }) => {
  const canvas = within(canvasElement);
  const textbox = canvas.getByRole('textbox');
  await userEvent.click(textbox);
  const button = await canvas.findByRole('button');
  setTimeout(async () => {
    await userEvent.click(button);
    expect(
      canvas.getByText(
        '* Comment is empty or too short. Min. length of the text should be 25 characters.'
      )
    ).toBeInTheDocument();
  }, 500);
};

export const CanCreateComment = CreateCommentTemplate.bind({});
CanCreateComment.args = CreateCommentExample.args;
CanCreateComment.play = async ({ canvasElement, args }) => {
  const canvas = within(canvasElement);
  const textbox = canvas.getByRole('textbox');
  await userEvent.click(textbox);
  await userEvent.type(textbox, 'This is a longer and more awesome comment!');
  const sendCommentBtn = await canvas.findByRole('button', {
    name: 'Send Comment',
  });
  await userEvent.click(sendCommentBtn);
  expect(args.setComments).toHaveBeenCalled();
};

export const CommentTooLong = CreateCommentTemplate.bind({});
const tooLongComment =
  'Lorem ipsum dolor sit amet consectetur adipisicing elit. Non enim minima velit molestiae voluptates deserunt aut perferendis aliquid? Ab sequi nesciunt possimus reprehenderit dolorum rerum assumenda consequuntur repellat quod aliquid! Lorem ipsum dolor sit, amet consectetur adipisicing elit. Dignissimos fugit, enim nulla commodi adipisci, beatae officia repudiandae esse ab obcaecati fugiat laborum deserunt numquam praesentium ex rerum sequi expedita unde.';
CommentTooLong.args = CreateCommentExample.args;
CommentTooLong.play = async ({ canvasElement }) => {
  const canvas = within(canvasElement);
  const textbox = canvas.getByRole('textbox');
  await userEvent.type(textbox, tooLongComment);
  await userEvent.click(textbox);
  const button = await canvas.findByRole('button');
  setTimeout(async () => {
    await userEvent.click(button);
    expect(
      canvas.getByText('* Comment is too long. Maximum is 250 characters.')
    ).toBeInTheDocument();
  }, 500);
};

export default {
  title: 'Molecules/Create comment',
  component: CreateComment,
};
