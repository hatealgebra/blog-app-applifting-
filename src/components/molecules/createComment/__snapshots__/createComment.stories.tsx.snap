// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Molecules/Create comment CreateCommentDisabled smoke-test 1`] = `
<form class="sc-kgvdhe jBqxex">
  <img class="sc-kDnyCx dyuMnk avatar"
       src="static/media/src/images/avatar-fallback.png"
       alt="fallback user avatar"
  >
  <textarea disabled
            placeholder="Join the discussion"
            rows="1"
            name="comment"
            class="sc-eDnVMP fmxlqA"
  >
  </textarea>
  <span class="sc-ipMvLY fOtdNV">
  </span>
</form>
`;

exports[`Molecules/Create comment CreateCommmentAuthorized smoke-test 1`] = `
<form class="sc-kgvdhe jBqxex">
  <img class="sc-kDnyCx dyuMnk avatar"
       src="static/media/src/images/avatar-fallback.png"
       alt="fallback user avatar"
  >
  <textarea placeholder="Join the discussion"
            rows="1"
            name="comment"
            class="sc-eDnVMP fmxlqA"
  >
  </textarea>
  <span class="sc-ipMvLY fOtdNV">
  </span>
</form>
`;
